crates/turborepo-cache/src/cache_archive/restore.rs

```diff
@@ -0,0 +1,1099 @@
+
+            let source_node = *nodes
+                .entry(processed_sourcename.clone())
+                .or_insert_with(|| graph.add_node(processed_sourcename.clone()));
+            let link_node = *nodes
+                .entry(processed_linkname.clone())
+                .or_insert_with(|| graph.add_node(processed_linkname.clone()));
+
+            graph.add_edge(source_node, link_node, ());
+
+            header_lookup.insert(processed_sourcename, entry.header().clone());
+        }
+
+        let nodes = petgraph::algo::toposort(&graph, None)
+            .map_err(|_| CacheError::CycleDetected(Backtrace::capture()))?;
+
+        for node in nodes {
+            let key = &graph[node];
+
+            let Some(header) = header_lookup.get(key) else {
+                continue
+            };
+            let file = restore_symlink_with_missing_target(anchor, header)?;
+            restored.push(file);
+        }
+
+        Ok(restored)
+    }
+}
+
+fn restore_entry<T: Read>(
+    anchor: &AbsoluteSystemPath,
+    entry: &mut Entry<T>,
+) -> Result<AnchoredSystemPathBuf, CacheError> {
+    let header = entry.header();
+
+    match header.entry_type() {
+        tar::EntryType::Directory => restore_directory(anchor, entry.header()),
+        tar::EntryType::Regular => restore_regular(anchor, entry),
+        tar::EntryType::Symlink => restore_symlink(anchor, entry.header()),
+        ty => Err(CacheError::UnsupportedFileType(ty, Backtrace::capture())),
+    }
+}
+
+pub fn canonicalize_name(name: &Path) -> Result<AnchoredSystemPathBuf, CacheError> {
+    #[allow(unused_variables)]
+    let PathValidation {
+        well_formed,
+        windows_safe,
+    } = check_name(name);
+
+    if !well_formed {
+        return Err(CacheError::MalformedName(
+            name.to_string_lossy().to_string(),
+            Backtrace::capture(),
+        ));
+    }
+
+    #[cfg(windows)]
+    {
+        if !windows_safe {
+            return Err(CacheError::WindowsUnsafeName(
+                name.to_string(),
+                Backtrace::capture(),
+            ));
+        }
+    }
+
+    // There's no easier way to remove trailing slashes in Rust
+    // because `OsString`s are really just `Vec<u8>`s.
+    let no_trailing_slash: PathBuf = name.components().collect();
+
+    // We know this is indeed anchored because of `check_name`,
+    // and it is indeed system because we just split and combined with the
+    // system path separator above
+    Ok(unsafe { AnchoredSystemPathBuf::new_unchecked(no_trailing_slash) })
+}
+
+#[derive(Debug, PartialEq)]
+struct PathValidation {
+    well_formed: bool,
+    windows_safe: bool,
+}
+
+fn check_name(name: &Path) -> PathValidation {
+    if name.as_os_str().is_empty() {
+        return PathValidation {
+            well_formed: false,
+            windows_safe: false,
+        };
+    }
+
+    let mut well_formed = true;
+    let mut windows_safe = true;
+    let name = name.to_string_lossy();
```

Comment:
We can't turn an `OsString` into a `Vec<u8>` easily because on Windows `OsString` is modeled as `Vec<u16>` whereas on Unix it's `Vec<u8>` (it's modeled as such, but actually it's stored as a custom superset of UTF-8 internally). We could in theory use a bit of unsafe to transmute an `OsString` to a `Vec<u8>` but that's relying on an internal, unstable API, which is probably not the greatest plan.